### PR TITLE
125 - Endpoint for calendar view

### PR DESCRIPTION
**Pull request by brownbearsoftware**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/CalendarTransformer.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            length = it.startDate.getDaysUntilInclusive(it.endDate).size,
+            startDate = it.startDate,
+            endDate = it.endDate,
+          )
+        }
+      }.toMutableList()
+
+      var startOfOpenPeriod: LocalDate? = null
+      var endOfOpenPeriod: LocalDate? = null
+      startDate.getDaysUntilInclusive(endDate).forEach { dateInRange ->
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/CalendarTransformer.kt:42]: I'm finding this block of code a little hard to follow, I'm wondering if we can do it in a different way by getting all the dates for the booked periods (Not sure if this is valid Kotlin!):

```kotlin
val bookedDates = transformedSchedule.map { it.startDate.getDaysUntilInclusive(it.endDate) }
val entirePeriod = startDate.getDaysUntilInclusive(endDate)

// Get all the missing dates in the schedule
val openDates = entirePeriod.minus(bookedDates)

openDates.forEach { i, startDate -> 
  val endDate: LocalDate? = null

  // Loop through all the dates following the current date
  openDates.drop(i + 1).forEach { length, date -> 
    // If the date is not the next day, we set next day to the previous day in the loop
    if (endDate !== startDate.addDays(length + 1)) {
      endDate = openDates[i -1]
    }
  }

 transformedSchedule += BedOccupancyOpenEntry(
   type = BedOccupancyEntryType.open,
   length = startDate!!.getDaysUntilInclusive(endDate!!).size,
   startDate = startDate!!,
   endDate = endDate!!,
 )
}
```

**Review by pezholio — COMMENTED** [review]
Looks great on the whole, just some suggestions to make the code a bit easier to grok

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/CalendarRepository.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      bookingsForPremisesQuery,
+      params,
+      ResultSetExtractor { resultSet ->
+        while (resultSet.next()) {
+          val bedKey = CalendarBedInfo(
+            bedId = UUID.fromString(resultSet.getString("bed_id")),
+            bedName = resultSet.getString("bed_name"),
+          )
+
+          ensureResultsMapContainsBed(bedKey)
```
[the remark this answers is not in the document]
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/repository/CalendarRepository.kt:66]: Unfortunately there's no ternary operator in Kotlin, closest would probably be:

```
resultsMap[bedKey] = if (resultsMap[bedKey] == null) resultsMap[bedKey]!! else mutableListOf()
```

But it doesn't feel particularly idiomatic - this did prompt me to remember that putIfAbsent is a thing though so I've updated to use that

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/CalendarTransformer.kt (hunk 1 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/CalendarTransformer.kt:42]: Wouldn't this add multiple entries for the same open period with the start date moving up by a day for each entry?

**Comment by brownbearsoftware** [comment]
[API endpoint for calendar view](https://trello.com/c/lIBoyJVa/125-api-endpoint-for-calendar-view)

**Review by pezholio — APPROVED** [review]
